### ISSUE TITLE
Making nonlinear predictions with Probabilistic Layers (DenseVariational)

### ISSUE THREAD

**Issue by pierremtb** [tensorflow/probability#815]
I've been playing the code of the [blog post](https://blog.tensorflow.org/2019/03/regression-with-probabilistic-layers-in.html) on Probabilistic Layers regression, trying to make it fit nonlinear stuff. I've stacked up two DenseVariational layers on top of the (mean, variance) one, as shown in this sample code. But no matter how much I play with the hyperparameters like the kl_weights, the learning rate of the coefficient in the softplus'ed variance, I can't get any good result. Would anyone care to tell me if I'm in the right direction? Thanks in advance!
```python
import numpy as np
import tensorflow as tf
import tensorflow_probability as tfp
import matplotlib.pyplot as plt

tfk = tf.keras
tfd = tfp.distributions

#%% HPs
N_tst = 100
N = 32

soft_0 = .01
lr = 0.01
klw = 1/N
epochs = 5000

dtype = "float64"
tfk.backend.set_floatx(dtype)

#%% Datagen
x_tst = np.linspace(-6, 6, N_tst).reshape(-1, 1)
y_tst = x_tst**3
lb = int(2/(2*6) * N_tst)
ub = int((2+2*4)/(2*6) * N_tst)
idx = np.random.choice(x_tst[lb:ub].shape[0], N, replace=False)
x = x_tst[lb + idx]
y = y_tst[lb + idx]
noise_std = 5
y = y + noise_std*np.random.randn(y.shape[0], y.shape[1])

#%% Prior and posterior
def prior_trainable(kernel_size, bias_size=0, dtype=None):
  n = kernel_size + bias_size
  return tf.keras.Sequential([
      tfp.layers.VariableLayer(n, dtype=dtype),
      tfp.layers.DistributionLambda(lambda t: tfd.Independent(
          tfd.Normal(loc=t, scale=1),
          reinterpreted_batch_ndims=1)),
  ])

def posterior_mean_field(kernel_size, bias_size=0, dtype=None):
  n = kernel_size + bias_size
  c = np.log(np.expm1(1.))
  return tf.keras.Sequential([
      tfp.layers.VariableLayer(2 * n, dtype=dtype),
      tfp.layers.DistributionLambda(lambda t: tfd.Independent(
          tfd.Normal(loc=t[..., :n],
                     scale=1e-5 + tf.nn.softplus(c + t[..., n:])),
          reinterpreted_batch_ndims=1)),
  ])

#%% Model definition
model = tfk.models.Sequential([
    tfp.layers.DenseVariational(20, activation="relu", input_shape=[1],
        make_posterior_fn=posterior_mean_field,
        make_prior_fn=prior_trainable,
        kl_weight=klw,
        dtype=dtype,
    ),
    tfp.layers.DenseVariational(20, activation="relu",
        make_posterior_fn=posterior_mean_field,
        make_prior_fn=prior_trainable,
        kl_weight=klw,
        dtype=dtype,
    ),
    tfp.layers.DenseVariational(1 + 1, activation="linear",
        make_posterior_fn=posterior_mean_field,
        make_prior_fn=prior_trainable,
        kl_weight=klw, dtype=dtype,
    ),
    tfp.layers.DistributionLambda(
        lambda t: tfd.Normal(loc=t[..., :1],
                            scale=1e-5 + tf.math.softplus(soft_0 * t[..., 1:]))),
])
model.compile(loss=lambda y, yhat: -yhat.log_prob(y),
              optimizer=tfk.optimizers.Adam(lr))
model.summary()

#%% Model training
model.fit(x, y, epochs=epochs)

#%% Predictions and plotting
y_pred_list = []

for i in range(100):
    y_dist = model(x_tst)
    y_pred_list.append(y_dist.mean().numpy())

y_preds = np.concatenate(y_pred_list, axis=1)

y_pred = np.mean(y_preds, axis=1)
y_pred_sig = np.std(y_preds, axis=1)
lower = y_pred - 2 * y_pred_sig
upper = y_pred + 2 * y_pred_sig

plt.fill_between(x_tst.ravel(), upper, lower,
                 facecolor='C0', alpha=0.3)
plt.plot(x_tst, y_pred)
plt.scatter(x, y, c="r")
plt.plot(x_tst, y_tst, "r--")
plt.show()
```
```console
Model: "sequential_15"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
dense_variational_45 (DenseV (None, 20)                120       
_________________________________________________________________
dense_variational_46 (DenseV (None, 20)                1260      
_________________________________________________________________
dense_variational_47 (DenseV (None, 2)                 126       
_________________________________________________________________
distribution_lambda_15 (Dist ((None, 1), (None, 1))    0         
=================================================================
Total params: 1,506
Trainable params: 1,506
Non-trainable params: 0
```

![Figure_1](https://user-images.githubusercontent.com/10795683/75378349-830f0500-58a1-11ea-8dce-163bc0fd38b1.png)

**Comment by wb815** [comment]
The ability of DenseVariational to fit nonlinear relationships seems to be *very* dependent on the scaling between the loc and scale parameters of the Gaussians in the posterior mean field. In my experience, reducing the scale by a factor of 10 to 100 has a nice regularization effect on the parameters fitted by the VariableLayers. As a quick demonstration, here's the impact of changing the definition of your posterior mean field to the following.

    def posterior_mean_field(kernel_size, bias_size=0, dtype=None):
      n = kernel_size + bias_size
      c = np.log(np.expm1(1.))
      return tf.keras.Sequential([
          tfp.layers.VariableLayer(2 * n, dtype=dtype),
          tfp.layers.DistributionLambda(lambda t: tfd.Independent(
              tfd.Normal(loc=t[..., :n],
                         scale=1e-5 + 0.02*tf.nn.softplus(c + t[..., n:])),
              reinterpreted_batch_ndims=1)),
      ])

![nonlin_readjusted](https://user-images.githubusercontent.com/37509266/75806714-1ea9e500-5d7c-11ea-8265-ea8698e46d5f.png)

**Comment by pierremtb** [comment]
That's indeed much better, thank you @wb815. I guess I'll close this then.
